### PR TITLE
CM-42771 - Add support of `.gitignore` files for a file excluding from scans

### PR DESCRIPTION
**Pull request by MarshalX**

### CODE REVIEW

File: cycode/cli/files_collector/walk_ignore.py [hunk 1]
```diff
@@ -0,0 +1,78 @@
...
+
+    if path:
+        yield path  # Include the top-level directory
+
+
+def _collect_top_level_ignore_files(path: str) -> List[str]:
+    ignore_files = []
+    for dir_path in _walk_to_top(path):
+        for ignore_file in _SUPPORTED_IGNORE_PATTERN_FILES:
+            ignore_file_path = os.path.join(dir_path, ignore_file)
```
[the remark this answers is not in the document]
Comment (MarshalX) [reply cycode/cli/files_collector/walk_ignore.py:31]: #cycode_sast_false_positive expected behavior; the input comes from the command argument; could be relative

[the remark this answers is not in the document]
Comment (MarshalX) [reply cycode/cli/files_collector/walk_ignore.py:31]: #cycode_sast_ignore_here

File: tests/cli/files_collector/test_walk_ignore.py [hunk 2]
```diff
@@ -0,0 +1,164 @@
...
+    result = list(_walk_to_top(path))
+    assert result == [
+        normpath('/a/b/c/d/e/f/g'),
+        normpath('/a/b/c/d/e/f'),
+        normpath('/a/b/c/d/e'),
+        normpath('/a/b/c/d'),
+        normpath('/a/b/c'),
+        normpath('/a/b'),
+        normpath('/a'),
+        normpath('/'),
```
Comment (elsapet) [inline tests/cli/files_collector/test_walk_ignore.py:30]: It's pretty 😍

**Review by elsapet — APPROVED** [review]

File: cycode/cli/files_collector/walk_ignore.py [hunk 3]
```diff
@@ -0,0 +1,42 @@
...
+            if os.path.exists(ignore_file_path):
```
[the remark this answers is not in the document]
Comment (MarshalX) [reply cycode/cli/files_collector/walk_ignore.py:30]: #cycode_sast_false_positive expected behavior; the input comes from the command argument; could be relative

File: cycode/cli/files_collector/walk_ignore.py [hunk 4]
```diff
@@ -0,0 +1,42 @@
...
+                logger.debug('Apply top level ignore file: %s', ignore_file_path)
+                ignore_files.append(ignore_file_path)
```
[the remark this answers is not in the document]
Comment (MarshalX) [reply cycode/cli/files_collector/walk_ignore.py:32]: #cycode_sast_false_positive expected behavior; the input comes from the command argument; could be relative

File: cycode/cli/utils/ignore_utils.py [hunk 5]
```diff
@@ -0,0 +1,459 @@
...
+            rel_dirpath = '' if dirpath == self.path else os.path.relpath(dirpath, self.path)
+
+            # decrease recursion depth of os.walk() by ignoring subdirectories because of topdown=True
+            # slicing ([:]) is mandatory to change dict in-place!
+            dirnames[:] = [
+                dirname for dirname in dirnames if not self.is_ignored(os.path.join(rel_dirpath, dirname, ''))
+            ]
+
+            # remove ignored files
+            filenames = [os.path.basename(f) for f in filenames if not self.is_ignored(os.path.join(rel_dirpath, f))]
```
[the remark this answers is not in the document]
Comment (MarshalX) [reply cycode/cli/utils/ignore_utils.py:404]: #cycode_sast_false_positive expected behavior; the input comes from the command argument; could be relative

File: cycode/cli/utils/ignore_utils.py [hunk 6]
```diff
@@ -0,0 +1,459 @@
...
+        """A wrapper for os.walk() without ignored files and subdirectories.
+        kwargs are passed to walk()."""
+
+        for dirpath, dirnames, filenames in os.walk(self.path, topdown=True, **kwargs):
+            rel_dirpath = '' if dirpath == self.path else os.path.relpath(dirpath, self.path)
+
+            # decrease recursion depth of os.walk() by ignoring subdirectories because of topdown=True
+            # slicing ([:]) is mandatory to change dict in-place!
+            dirnames[:] = [
+                dirname for dirname in dirnames if not self.is_ignored(os.path.join(rel_dirpath, dirname, ''))
```
[the remark this answers is not in the document]
Comment (MarshalX) [reply cycode/cli/utils/ignore_utils.py:400]: #cycode_sast_false_positive expected behavior; the input comes from the command argument; could be relative

File: cycode/cli/utils/ignore_utils.py [hunk 7]
```diff
@@ -0,0 +1,459 @@
...
+    def _load_path(self, path: str) -> Optional[IgnoreFilter]:
+        try:
+            return self._path_filters[path]
+        except KeyError:
+            pass
+
+        if not self._ignore_file_name:
+            self._path_filters[path] = None
+        else:
+            p = os.path.join(self._top_path, path, self._ignore_file_name)
```
[the remark this answers is not in the document]
Comment (MarshalX) [reply cycode/cli/utils/ignore_utils.py:337]: #cycode_sast_false_positive expected behavior; the input comes from the command argument; could be relative

**Comment by MarshalX** [comment]
#cycode_unfail_me

**Review by elsapet — APPROVED** [review]

**Pull request merged**